### ISSUE TITLE
std::generator usage with Clang causes warning/error when compiling HPX with C++23 and coroutines enabled

### ISSUE THREAD

**Issue by Prachethan7** [llvm/llvm-project#136607]
I'm encountering a compiler warning (treated as error) when compiling [HPX](https://github.com/STEllAR-GROUP/hpx) with -std=c++23 and coroutine support using Clang 17.0.6.

The error is related to the usage of std::generator and seems to stem from header processing logic.

```
cmake .. \
  -DHPX_DIR=/home/neo/packages/hpx/install/lib/cmake/HPX \
  -DHPX_WITH_CXX_STANDARD=23 \
  -DHPX_HAVE_CXX23_STD_GENERATOR=ON \
  -DHPX_WITH_CXX20_COROUTINES=ON

cmake --build . --target tests

```

Clang version:
```
Debian clang version 19.1.7 (3)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-19/bin

```

The error disappears when adding a missing ;, but this raises a concern: the warning may indicate a deeper issue in how std::generator is parsed or supported.

Would it be possible to:

    Confirm whether this is an expected behavior?

    Clarify support status for std::generator with Clang 17?

    Possibly improve diagnostic feedback or preprocessor logic?

[compiler_crash.log](https://github.com/user-attachments/files/19840012/compiler_crash.log)

**Comment by EugeneZelenko** [comment]
Could you please try 20 or `main` branch? https://godbolt.org should be helpful.

**Comment by Prachethan7** [comment]
> Could you please try 20 or `main` branch? https://godbolt.org should be helpful.

Thank you for the quick response and for referring me to godbolt.org — I’ll definitely take a closer look using that.

Just to clarify, the code I’m working with specifically requires C++23 due to the use of std::generator, so trying with C++20 or switching branches might not reflect the same issue.

**Comment by Endilll** [comment]
> Just to clarify, the code I’m working with specifically requires C++23 due to the use of std::generator, so trying with C++20 or switching branches might not reflect the same issue.

Eugene was referring to Clang 20, not C++20 mode. Clang 17 has been long out of support, so we need to know whether your problem is reproducible with the latest release (20) or in-development version (21).

**Comment by Prachethan7** [comment]
> > Just to clarify, the code I’m working with specifically requires C++23 due to the use of std::generator, so trying with C++20 or switching branches might not reflect the same issue.
> 
> Eugene was referring to Clang 20, not C++20 mode. Clang 17 has been long out of support, so we need to know whether your problem is reproducible with the latest release (20) or in-development version (21).


Thanks for the clarification Endilll . I've used the Clang version of 19.1.7
